lib/pure/hashes.nim

```diff
@@ -136,6 +139,17 @@ proc hash*[T: Ordinal](x: T): Hash {.inline.} =
   ## Efficient hashing of other ordinal types (e.g. enums).
   result = ord(x)
 
+template multibyteHashImpl(result: Hash, x: typed, start, stop: int) =
```

Comment:
Here are some results on my machine (i7-970) for some basic tests I did. These are average times, but there is still variance so single-digit percent difference should be taken as zero.

test | old algo (ms) | new algo (ms) | difference
--- | --- | --- | ---
10_000 English words | 0.731 | 0.750 | +2.6%
20_000 English words | 1.517 | 1.563 | +3.0%
50_000 English words | 4.559 | 4.678 | +2.6%
--- | --- | --- | ---
20_000 random 4-char strings | 0.497 | 0.503 | +1.2%
20_000 random 8-char strings | 0.775 | 1.077 | **+38.9%** ??
20_000 random 9-char strings | 0.867 | 0.614 | -29.1%
20_000 random 12-char strings | 0.952 | 0.687 | -27.8%
20_000 random 16-char strings | 1.162 | 0.592 | -49.0%
20_000 random 24-char strings | 1.504 | 0.600 | -60.1%
20_000 random 32-char strings | 1.860 | 0.656 | -64.7%
20_000 random 64-char strings | 3.274 | 0.851 | -74.0%

As the length of string/array/seq increases, so does the performance of the new algorithm.

The only exception is length of 8, which is unexpectedly slow (the first while loop is called once, the second one is called zero times). Increasing the length to 9 (first loop once, second once) has a speed-up which would be expected for length of 8 too.

I've tried several different ways to write the first loop, but all had the similar results.

@mratsim do you maybe know what is going on here? Any tips?

Reply:
Can you share your benchmarking script?

Reply:
> Can you share your benchmarking script?

It was just a quick/simple "hash a word, add a hash to set, do something with that set (don't time it) just in case", repeat that dozens of time, and report the average:

```nim
proc hashTesting(wordlist: seq[string], f: proc (x: string): Hash, name: string): float =
  let
    amount = 20_000
    repetitions = 48
  var
    collisions = newSeq[int](repetitions)
    highest = newSeq[int](repetitions)
    times = newSeq[float](repetitions)

  for i in 0 ..< repetitions:
    let words = wordList[amount*i ..< amount*i+amount]
    var s = initHashSet[int](sets.rightSize(amount))
    var t = cpuTime()
    for w in words:
      s.incl f(w)
    t = cpuTime() - t
    times[i] = t
    collisions[i] = amount - s.len
    highest[i] = s.toSeq.max
  echo "--"
  echo name
  echo highest.min
  echo collisions.sum / collisions.len
  echo "max: ", times.max * 1000.0
  echo "min: ", times.min * 1000.0
  return times.sum / times.len.float * 1000.0
```

where `f: proc (x: string): Hash` is either the old one or the new one (or several others that I've tested).

Reply:
> The only exception is length of 8, which is unexpectedly slow

this is explained here: https://github.com/nim-lang/Nim/issues/11581